### PR TITLE
[AMDGPU] support 64-bit immediates in SIInstrInfo::FoldImmediate

### PR DESCRIPTION
**Pull request by rampitec**
This is a part of https://github.com/llvm/llvm-project/issues/67781. Until we select more 64-bit move immediates the impact is minimal.

### CODE REVIEW

**Review by Sisyph — APPROVED** [review]
LGTM, besides nit.

File: llvm/test/CodeGen/AMDGPU/packed-fp32.ll [hunk 1]
```diff
@@ -308,7 +308,7 @@ define amdgpu_kernel void @fmul_v2_v_v_splat(ptr addrspace(1) %a) {
 ; GCN-LABEL: {{^}}fmul_v2_v_lit_splat:
 ; GFX900-COUNT-2: v_mul_f32_e32 v{{[0-9]+}}, 4.0, v{{[0-9]+}}
 ; PACKED-SDAG:    v_pk_mul_f32 v[{{[0-9:]+}}], v[{{[0-9:]+}}], 4.0 op_sel_hi:[1,0]{{$}}
-; PACKED-GISEL:   v_pk_mul_f32 v[{{[0-9:]+}}], v[{{[0-9:]+}}], s[{{[0-9:]+}}]{{$}}
+; PACKED-GISEL:   v_pk_mul_f32 v[{{[0-9:]+}}], v[{{[0-9:]+}}], {{[sv]}}[{{[0-9:]+}}]{{$}}
```
[the remark this answers is not in the document]
Comment (rampitec) [reply llvm/test/CodeGen/AMDGPU/packed-fp32.ll]: Turns out this was not needed, this change went away somewhere in the middle of the implementation.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 2]
```diff
@@ -3220,19 +3219,45 @@ bool SIInstrInfo::FoldImmediate(MachineInstr &UseMI, MachineInstr &DefMI,
...
+    case AMDGPU::sub1:
+      return Hi_32(Imm);
+    case AMDGPU::lo16:
+      return APInt(16, Imm).getSExtValue();
+    case AMDGPU::hi16:
+      return APInt(32, Imm).ashr(16).getSExtValue();
+    case AMDGPU::sub1_lo16:
+      return APInt(16, Hi_32(Imm)).getSExtValue();
+    case AMDGPU::sub1_hi16:
+      return APInt(32, Hi_32(Imm)).ashr(16).getSExtValue();
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:3238]: Nit: seems over-engineered to use `APInt`. Simpler and faster to return `(int16_t)(Imm >> 48)`.

File: llvm/lib/Target/AMDGPU/SIInstructions.td [hunk 3]
```diff
@@ -151,6 +151,7 @@ def S_MOV_B64_IMM_PSEUDO : SPseudoInstSI <(outs SReg_64:$sdst),
   let SchedRW = [WriteSALU, Write64Bit];
   let Size = 16; // Needs maximum 2 s_mov_b32 instructions 8 byte long each.
   let Uses = [];
+  let UseNamedOperandTable = 1;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstructions.td:154]: What is this for?

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 2 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:3238]: int16_t is signed so the implicit conversion to the return type will sign extend it.

File: llvm/lib/Target/AMDGPU/SIInstructions.td (hunk 3 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIInstructions.td:154]: Makes sense, thanks. I did not even know there was a flag for this. I thought it happened automatically.

**Pull request merged**